### PR TITLE
ENHANCEMENT: Add new CacheProxy to segment versioned state

### PR DESCRIPTION
**Pull request by unclecheese**
Resolves: https://github.com/silverstripe/silverstripe-framework/issues/8020

### CODE REVIEW

File: cache/CacheProxy.php [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    public function clean($mode = Zend_Cache::CLEANING_MODE_ALL, $tags = array())
+    {
+        return $this->container->clean($mode, $tags);
+    }
+
+    /**
+     * Creates a dynamic key based on versioned state
+     * @param $key
```
Comment (robbieaverill) [inline cache/CacheProxy.php]: Missing type

File: cache/CacheProxy.php [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    /**
+     * @param array $directives
+     */
+    public function setDirectives($directives) {
+        $this->container->setDirectives($directives);
+    }
+
+    public function setConfig(Zend_Config $config)
```
[the remark this answers is not in the document]
Comment (unclecheese) [reply cache/CacheProxy.php]: It's optional. You can use it if you want to mark something explicitly documented, but this is standards compliant as it is.

[the remark this answers is not in the document]
Comment (unclecheese) [reply cache/CacheProxy.php]: In fact, I think I'll just remove all the others, because it's inconsistent.

File: docs/en/04_Changelogs/3.7.0.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
+# 3.7.0
+
+### Versioned cache segmentation
+
+The cache API now maintains separate cache pools for each versioned stage. This prevents users from caching draft data and then exposing it on the live stage.
```
Comment (chillu) [inline docs/en/04_Changelogs/3.7.0.md]: Call it `SS_Cache` API, to be clear about which cache you mean (e.g. not HTTP cache).

Also, I think the intent for this change should be more obvious. Suggested rewording:

> This prevents developers from caching draft data and then accidentally exposing it on the live stage without potentially required authorisation checks. Unless you rely on caching across stages, you don't need to change your own code for this change to take effect. Note that cache keys will be internally rewritten, causing any existing cache items to become invalid when this change is deployed.

File: docs/en/04_Changelogs/3.7.0.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+```php
+// Before:
+$cache = SS_Cache::factory('myapp');
+Versioned::set_reading_mode('Stage.Live');
```
Comment (chillu) [inline docs/en/04_Changelogs/3.7.0.md]: Shouldn't the reading mode be the other way around, first `Stage.Stage`, then `Stage.Live`?

**Comment by chillu** [comment]
OK I'm happy with this, but would like a second opinion either from @dhensby or @sminnee

**Comment by kinglozzer** [comment]
This captures all the methods defined in `Zend_Cache_Core`, but the `Zend_Cache_Frontend_<driver>` classes sometimes add extra methods that wouldn’t be covered by this. Could/should we add `__call()` for those cases? I imagine they’re scarcely used, but would be broken by this change

**Comment by chillu** [comment]
@unclecheese Can you please go through the 3.x and 4.x default caches in core, and opt out where this behaviour isn't required? Otherwise we'll cause performance regressions, e.g. for inherited permissions caches. See Damian's comment on https://github.com/silverstripe/silverstripe-versioned/pull/148

**Comment by unclecheese** [comment]
Sure. The 4.x ones are already done, though. Do you need any changes to that? https://github.com/silverstripe/silverstripe-framework/pull/8093#issuecomment-393313829

File: docs/en/04_Changelogs/3.7.0.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
+# 3.7.0
+
+### Versioned cache segmentation
```
Comment (chillu) [inline docs/en/04_Changelogs/3.7.0.md]: @unclecheese Can you please document this in a persistent place in the docs as well? Also apply this to the 4.x PRs

**Comment by chillu** [comment]
@unclecheese Can you please check the Travis failures? E.g. there's PHP 5.3 incompatible syntax.

File: docs/en/04_Changelogs/3.7.0.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+$cache = SS_Cache::factory('myapp');
+Versioned::set_reading_mode('Stage.Stage');
+$cache->save('Some draft content. Not for public viewing yet.', 'my_key');
+Versioned::set_reading_mode('Stage.Live');
+$cache->load('my_key'); // null
+```
+Data that is not content sensitive can be cached across stages by simply opting out of the segmented cache with the `disable-container` argument.
+
+```php
+$cache = SS_Cache::factory('myapp', 'Output', array('disable-container' => true));
```
Comment (chillu) [inline docs/en/04_Changelogs/3.7.0.md]: Should be `disable-segmentation`

**Comment by chillu** [comment]
Have you checked how this affects `SSViewer` partial caching? Presumably this is also using `SS_Cache::factory()`, and now includes the stage twice in the cache key? Once from the `SS_Cache::factory()` logic, once from the custom `SSViewer` logic? That unnecessarily changes existing caches, and feels a bit confusing.

**Comment by tractorcow** [comment]
Yes we can use the same disable-segmentation key in SSViewer to disable that. It's not necessary to support it where we already respect viewing mode.

**Comment by tractorcow** [comment]
@ingo addressed above.

**Comment by tractorcow** [comment]
Behat issues fixed over at https://github.com/silverstripe/silverstripe-behat-extension/pull/178

**Comment by tractorcow** [comment]
Testing this PR locally with that behat PR looks good. Just need to get it running on travis.

**Comment by robbieaverill** [comment]
PR to increase memory limit in Travis: https://github.com/silverstripe/silverstripe-framework/pull/8129

**Comment by chillu** [comment]
@unclecheese Behat tests are still failing (just retriggered it a few mins ago), even though the memory usage PR Robbie referenced above got merged. Maybe we need to have a release for behat-extension?

File: docs/en/04_Changelogs/3.7.0.md (hunk 5 above)
Comment (chillu) [reply docs/en/04_Changelogs/3.7.0.md]: @unclecheese Still outstanding

**Comment by tractorcow** [comment]
just tagged behat extension again. Let's see if it fixes it.

**Comment by tractorcow** [comment]
Only remaining issue is a postgres issue that seems unrelated to this issue.

**Pull request merged**